### PR TITLE
Enhance Security, Validation, and Error Handling with CORS Support for Swagger

### PR DESCRIPTION
**Pull request by RYelzhan**
This pull request includes several key updates and improvements to the project:

1) CORS and Swagger Editor Support:
- Enabled CORS for the Swagger Editor URL to allow cross-origin requests.
- Fixed an issue with annotation inheritance from interfaces, specifically for custom annotations.

2) Validation Improvements:
- Enhanced DTO argument validation in controller methods by implementing custom Jakarta Bean Validators. These validators handle specific cases, such as validating TrainingType, TraineeUsername, and TrainerUsername.

3) Swagger Annotation Refactor:
- Moved Swagger annotations from concrete controller methods to interface methods to ensure better maintainability and consistency.

4) Error Handling Improvements:
- Improved error handling by making error messages visible to end users, enhancing the application's usability.
- Updated OpenAPI documentation annotations to reflect recent changes.

5) Security Enhancements:
- Added brute force protection to the authentication system.
- Enabled JWT token-based authentication and replaced manual authentication checks with Spring Security.
- Implemented password hashing and fixed issues with password comparison in hashed form.
- Added graceful exception handling for incorrect token entries.

6) Miscellaneous Improvements:
- Added a health indicator to check the initialization state of the database.
- Refactored code for better organization and maintainability.

This pull request improves the overall security, robustness, and user experience of the application while ensuring better documentation and validation practices.

### CODE REVIEW

File: src/main/java/com/epam/wca/gym/service/impl/LoginAttemptService.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+        if (attempts == MAX_ATTEMPT) {
+            blockTime.put(key, new Date(System.currentTimeMillis() + BLOCK_TIME_MILLIS));
+        }
+    }
+
+    public boolean isBlocked() {
+        String clientIp = getClientIP();
+
+        int attempts = attemptsMap.getOrDefault(clientIp, 0);
```
Comment (MadinaBaizhanova) [inline src/main/java/com/epam/wca/gym/service/impl/LoginAttemptService.java:42]: I would replace all these zero-s in the class by meaningful constants 🤓

File: src/main/java/com/epam/wca/gym/service/impl/LoginAttemptService.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+import java.util.Date;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+@Getter
+@Service
+public class LoginAttemptService {
+    private static final int MAX_ATTEMPT = 3;
+    private static final int BLOCK_TIME_MILLIS = 300000;
```
Comment (MadinaBaizhanova) [inline src/main/java/com/epam/wca/gym/service/impl/LoginAttemptService.java:18]: I would write `300_000` just for readability, but this is something super minor.

File: src/main/java/com/epam/wca/gym/config/SecurityConfig.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        authProvider.setPasswordEncoder(passwordEncoder());
+
+        return authProvider;
+    }
+
+    // used by my provider
+    @Bean
+    public PasswordEncoder passwordEncoder() {
+        // default strength (a.k.a hash rounds) is 10
+        return new BCryptPasswordEncoder(11);
```
Comment (MadinaBaizhanova) [inline src/main/java/com/epam/wca/gym/config/SecurityConfig.java:96]: I would replace this number by a meaningful constant and would remove the comments above.

File: src/main/java/com/epam/wca/gym/aop/check/CheckTraineeAspect.java [hunk 4]
```diff
@@ -8,19 +8,25 @@
...
     public void checkTraineePointcut(HttpServletRequest request) {
         // This method is empty because it serves as a pointcut definition.
     }
 
     @Around(value = "checkTraineePointcut(request)", argNames = "pjp,request")
     public Object checkTrainee(ProceedingJoinPoint pjp, HttpServletRequest request) throws Throwable {
-        User authenticatedUser = (User) request.getAttribute("authenticatedUser");
+        User authenticatedUser = (User) request.getAttribute(authenticatedUserRequestAttributeName);
+
+        System.out.println(authenticatedUser);
```
Comment (MadinaBaizhanova) [inline src/main/java/com/epam/wca/gym/aop/check/CheckTraineeAspect.java:29]: Consider removing this sout statement or replacing it by a logger. I have once heard that it is not recommended to use them.

File: src/main/java/com/epam/wca/gym/aop/check/CheckTrainee.java
Comment (MadinaBaizhanova) [inline src/main/java/com/epam/wca/gym/aop/check/CheckTrainee.java:1]: Consider adding some javadoc comments to the CheckTrainee and CheckTrainer annotations to explain why they are needed, how they work, etc.

File: src/main/java/com/epam/wca/gym/advice/AuthenticationFailureListener.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    public void onApplicationEvent(@NonNull AuthenticationFailureBadCredentialsEvent e) {
+        HttpServletRequest request = getCurrentHttpRequest();
+        if (request != null) {
+            final String xfHeader = request.getHeader("X-Forwarded-For");
+            String clientIp;
+
+            if (xfHeader == null || xfHeader.isEmpty() || !xfHeader.contains(request.getRemoteAddr())) {
+                clientIp = request.getRemoteAddr();
+            } else {
+                clientIp = xfHeader.split(",")[0];
```
[the remark this answers is not in the document]
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/advice/AuthenticationFailureListener.java:30]: Done.

File: src/main/java/com/epam/wca/gym/advice/AuthenticationFailureListener.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    private final LoginAttemptService loginAttemptService;
+
+    @Override
+    public void onApplicationEvent(@NonNull AuthenticationFailureBadCredentialsEvent e) {
+        HttpServletRequest request = getCurrentHttpRequest();
+        if (request != null) {
+            final String xfHeader = request.getHeader("X-Forwarded-For");
+            String clientIp;
+
+            if (xfHeader == null || xfHeader.isEmpty() || !xfHeader.contains(request.getRemoteAddr())) {
```
[the remark this answers is not in the document]
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/advice/AuthenticationFailureListener.java:27]: Done.

File: src/main/java/com/epam/wca/gym/controller/AuthenticationController.java [hunk 7]
```diff
@@ -81,7 +81,5 @@ public UserAuthenticatedDTO registerTrainee(@RequestBody @Valid TraineeRegistrat
     )
     @PostMapping(value = "/register/trainer")
```
[the remark this answers is not in the document]
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/controller/AuthenticationController.java:82]: Done.

File: src/main/java/com/epam/wca/gym/service/impl/LoginAttemptService.java (hunk 1 above)
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/service/impl/LoginAttemptService.java:42]: Okay okay, done.

File: src/main/java/com/epam/wca/gym/advice/AuthenticationFailureListener.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+            } else {
+                clientIp = xfHeader.split(",")[0];
+            }
+
+            loginAttemptService.loginFailed(clientIp);
+        }
+    }
+
+    private HttpServletRequest getCurrentHttpRequest() {
+        RequestAttributes requestAttributes = RequestContextHolder.getRequestAttributes();
```
[the remark this answers is not in the document]
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/advice/AuthenticationFailureListener.java:38]: Done.

File: src/main/java/com/epam/wca/gym/controller/TraineeController.java [hunk 9]
```diff
@@ -55,14 +40,8 @@ public class TraineeController {
...
-    @GetMapping("/profiles")
-    @CheckTrainee
-    @Transactional
-    public TraineeSendDTO getTraineeProfile(HttpServletRequest request) {
-        var trainee = (Trainee) request.getAttribute(authenticatedUserRequestAttributeName);
-
-        return DTOFactory.createTraineeSendDTO(trainee);
-    }
+    @GetMapping(value = "/profiles", consumes = MediaType.ALL_VALUE)
+    TraineeSendDTO getTraineeProfile(HttpServletRequest request);
```
[the remark this answers is not in the document]
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/controller/TraineeController.java:44]: Done, methods in other controllers are updated as well.

File: src/main/java/com/epam/wca/gym/controller/TraineeController.java [hunk 10]
```diff
@@ -80,16 +59,11 @@ public TraineeSendDTO getTraineeProfile(HttpServletRequest request) {
             responseCode = "401",
             description = ResponseMessages.UNAUTHORIZED_ACCESS_DESCRIPTION
     )
-    @PutMapping(value = "/profiles", consumes = MediaType.APPLICATION_JSON_VALUE)
-    @CheckTrainee
-    public TraineeSendDTO updateTraineeProfile(
+    @PutMapping(value = "/profiles")
+    TraineeSendDTO updateTraineeProfile(
```
[the remark this answers is not in the document]
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/controller/TraineeController.java:63]: Done.

File: src/main/java/com/epam/wca/gym/controller/TraineeController.java [hunk 11]
```diff
@@ -173,15 +131,10 @@ public List<TrainerBasicDTO> updateTrainerList(
             description = ResponseMessages.UNAUTHORIZED_ACCESS_DESCRIPTION
     )
     @GetMapping("/trainings/filter")
-    @CheckTrainee
-    public List<TrainingBasicDTO> getTraineeTrainingsList(
+    List<TrainingBasicDTO> getTraineeTrainingsList(
```
[the remark this answers is not in the document]
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/controller/TraineeController.java:134]: Done.

File: src/main/java/com/epam/wca/gym/controller/TraineeController.java [hunk 12]
```diff
@@ -200,16 +153,9 @@ public List<TrainingBasicDTO> getTraineeTrainingsList(
             description = ResponseMessages.UNAUTHORIZED_ACCESS_DESCRIPTION
     )
     @PostMapping("/trainings")
-    @CheckTrainee
     @ResponseStatus(HttpStatus.CREATED)
-    public String createTraineeTraining(
+    String createTraineeTraining(
```
[the remark this answers is not in the document]
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/controller/TraineeController.java:157]: Done.

File: src/main/java/com/epam/wca/gym/controller/TraineeController.java [hunk 13]
```diff
@@ -173,15 +131,10 @@ public List<TrainerBasicDTO> updateTrainerList(
             description = ResponseMessages.UNAUTHORIZED_ACCESS_DESCRIPTION
     )
     @GetMapping("/trainings/filter")
```
[the remark this answers is not in the document]
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/controller/TraineeController.java:133]: Done.

File: src/main/java/com/epam/wca/gym/controller/TraineeController.java [hunk 14]
```diff
@@ -149,17 +113,11 @@ public List<TrainerBasicDTO> getNotAssignedTrainers(HttpServletRequest request)
             description = ResponseMessages.UNAUTHORIZED_ACCESS_DESCRIPTION
     )
     @PutMapping(value = "/trainers/add", consumes = MediaType.APPLICATION_JSON_VALUE)
```
[the remark this answers is not in the document]
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/controller/TraineeController.java:115]: Thanks, done.

File: src/main/java/com/epam/wca/gym/controller/TraineeController.java [hunk 15]
```diff
@@ -1,48 +1,33 @@
...
-@RequestMapping(value = "/user/trainee")
-@RequiredArgsConstructor
-public class TraineeController {
-    private final TraineeService traineeService;
-
-    @Value("${gym.api.request.attribute.user}")
-    private String authenticatedUserRequestAttributeName;
-
+@SecurityRequirement(name = "jwtToken")
+@RequestMapping(value = "/user/trainee", consumes = MediaType.APPLICATION_JSON_VALUE)
```
[the remark this answers is not in the document]
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/controller/TraineeController.java:29]: Done.

File: src/main/java/com/epam/wca/gym/controller/TrainerController.java [hunk 16]
```diff
@@ -26,20 +20,12 @@
...
-@RequestMapping(value = "user/trainer")
-@RequiredArgsConstructor
-public class TrainerController {
-    private final TrainerService trainerService;
-
-    @Value("${gym.api.request.attribute.user}")
-    private String authenticatedUserRequestAttributeName;
-
+@SecurityRequirement(name = "jwtToken")
+@RequestMapping(value = "user/trainer", consumes = MediaType.APPLICATION_JSON_VALUE)
```
[the remark this answers is not in the document]
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/controller/TrainerController.java:27]: Done.

File: src/main/java/com/epam/wca/gym/controller/TrainingController.java [hunk 17]
```diff
@@ -1,25 +1,19 @@
...
 
-@SecurityRequirement(name = "basicAuth")
-@RestController
-@RequestMapping(value = "/training")
-@RequiredArgsConstructor
-public class TrainingController {
-    private final TrainingTypeService trainingTypeService;
-
+@SecurityRequirement(name = "jwtToken")
+@RequestMapping(value = "/training", consumes = MediaType.ALL_VALUE)
```
[the remark this answers is not in the document]
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/controller/TrainingController.java:15]: Done.

File: src/main/java/com/epam/wca/gym/controller/UserController.java [hunk 18]
```diff
@@ -2,33 +2,22 @@
...
-@RestController
-@RequestMapping(value = "/user")
-@RequiredArgsConstructor
-public class UserController {
-    private final UserService userService;
-    @Value("${gym.api.request.attribute.user}")
-    private String authenticatedUserRequestAttributeName;
 
+@SecurityRequirement(name = "jwtToken")
+@RequestMapping(value = "/user", consumes = MediaType.APPLICATION_JSON_VALUE)
```
[the remark this answers is not in the document]
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/controller/UserController.java:19]: Done.

File: src/main/java/com/epam/wca/gym/controller/UserController.java [hunk 19]
```diff
@@ -92,42 +69,9 @@ public String changeUserPassword(
             responseCode = "401",
             description = ResponseMessages.UNAUTHORIZED_ACCESS_DESCRIPTION
     )
-    @PatchMapping(value = "/change/active", consumes = MediaType.APPLICATION_JSON_VALUE)
-    public String activateDeactivateUser(
+    @PatchMapping(value = "/change/active")
```
[the remark this answers is not in the document]
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/controller/UserController.java:72]: Done.

File: src/main/java/com/epam/wca/gym/service/impl/LoginAttemptService.java (hunk 2 above)
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/service/impl/LoginAttemptService.java:18]: Done.

File: src/main/java/com/epam/wca/gym/config/SecurityConfig.java (hunk 3 above)
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/config/SecurityConfig.java:96]: Yeah, I left comments for myself in future to use during demo. Removed them. Done.

File: src/main/java/com/epam/wca/gym/aop/check/CheckTraineeAspect.java (hunk 4 above)
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/aop/check/CheckTraineeAspect.java:29]: Ooops. It should not have been there. Done.

File: src/main/java/com/epam/wca/gym/advice/RestAuthenticationEntryPoint.java [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+        if (authException.getCause() instanceof UserBlockedException userBlockedException) {
+            response.setStatus(HttpServletResponse.SC_FORBIDDEN);
+            errorMessage = userBlockedException.getMessage();
+        } else if (authException.getCause() instanceof UsernameNotFoundException usernameNotFoundException) {
+            response.setStatus(HttpServletResponse.SC_FORBIDDEN);
+            errorMessage = usernameNotFoundException.getMessage();
+        } else {
+            response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+            errorMessage = authException.getMessage();
+        }
```
Comment (maximelyashevich) [inline src/main/java/com/epam/wca/gym/advice/RestAuthenticationEntryPoint.java:33]: Would it be possible to eliminate the use of `if-else-if` solution?

File: src/main/java/com/epam/wca/gym/aop/check/CheckTrainerAspect.java [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+
+    @Around(value = "checkTrainerPointcut(request)", argNames = "pjp,request")
+    public Object checkTrainer(ProceedingJoinPoint pjp, HttpServletRequest request) throws Throwable {
+        User authenticatedUser = (User) request.getAttribute(authenticatedUserRequestAttributeName);
+
+        if (authenticatedUser instanceof Trainer) {
+            return pjp.proceed();
+        }
+        throw new ForbiddenActionException("You are not Trainer.");
+    }
```
Comment (maximelyashevich) [inline src/main/java/com/epam/wca/gym/aop/check/CheckTrainerAspect.java:56]: Shouldn't it be migrated to Spring Security?

File: src/main/java/com/epam/wca/gym/aop/validation/TraineeValidator.java [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+public class TraineeValidator implements ConstraintValidator<ValidTrainee, String> {
+    private final TraineeRepository traineeRepository;
+
+    @Override
+    public boolean isValid(String traineeUsername, ConstraintValidatorContext context) {
+        if (traineeUsername == null || traineeUsername.trim().isEmpty()) {
```
Comment (maximelyashevich) [inline src/main/java/com/epam/wca/gym/aop/validation/TraineeValidator.java:16]: Please extract to a private method this check.

File: src/main/java/com/epam/wca/gym/controller/AuthenticationController.java [hunk 23]
```diff
@@ -38,16 +28,28 @@ public class AuthenticationController {
...
+            description = "Logouts a user based on the username and password."
+    )
+    @ApiResponse(
+            responseCode = "200",
+            description = "Logout successful"
+    )
+    @ApiResponse(
+            responseCode = "401",
+            description = ResponseMessages.UNAUTHORIZED_ACCESS_DESCRIPTION
+    )
```
Comment (maximelyashevich) [inline src/main/java/com/epam/wca/gym/controller/AuthenticationController.java:49]: Can you please extract Open API annotations to a dedicated interfaces to keep your controllers clean an concise. Please consider this refactoring for all your controllers.

File: src/main/java/com/epam/wca/gym/filter/JwtAuthenticationFilter.java [hunk 24]
```diff
@@ -0,0 +1,126 @@
...
+@RequiredArgsConstructor
+public class JwtAuthenticationFilter extends OncePerRequestFilter {
+    private final JwtService jwtService;
+    private final UserDetailsService userDetailsService;
+    private static final Set<String> ALLOWED_PREFIXES = Set.of(
+            "/authentication",
+            "/h2-console",
+            "/admin",
+            "/v3/api-docs"
+    );
```
Comment (maximelyashevich) [inline src/main/java/com/epam/wca/gym/filter/JwtAuthenticationFilter.java:36]: Does it make sense to move your allowed-list configuration to `application.properties` to increase your project configurability?

File: src/main/java/com/epam/wca/gym/monitoring/DatabaseInitMonitor.java [hunk 25]
```diff
@@ -19,21 +20,44 @@ public class DatabaseInitMonitor implements HealthIndicator {
...
     public Health health() {
         if (isHealthGood()) {
-            return Health.up().withDetails(Map.of(DATABASE_INIT, "Database was correctly initialised")).build();
+            Map<String, Integer> details = new HashMap<>();
+
+            details.put("users_table", getTableRowCount("users"));
+            details.put("trainees_table", getTableRowCount("trainees"));
+            details.put("trainers_table", getTableRowCount("trainers"));
+            details.put("trainings_table", getTableRowCount("trainings"));
+            details.put("training_type_table", getTableRowCount("training_type"));
```
Comment (maximelyashevich) [inline src/main/java/com/epam/wca/gym/monitoring/DatabaseInitMonitor.java:29]: Can you please consider `Map.of()` from Java 9?

File: src/main/java/com/epam/wca/gym/monitoring/DatabaseInitMonitor.java [hunk 26]
```diff
@@ -19,21 +20,44 @@ public class DatabaseInitMonitor implements HealthIndicator {
...
     }
+
+    private Integer getTableRowCount(String tableName) {
+        String query = "SELECT COUNT(*) FROM %s".formatted(tableName);
+        try {
+            return jdbcTemplate.queryForObject(query, Integer.class);
+        } catch (Exception e) {
+            return -1;
+        }
+    }
```
Comment (maximelyashevich) [inline src/main/java/com/epam/wca/gym/monitoring/DatabaseInitMonitor.java:62]: Can you extract please your String values to constants?

File: src/main/java/com/epam/wca/gym/service/impl/LoginAttemptService.java [hunk 27]
```diff
@@ -0,0 +1,75 @@
...
+import java.util.Date;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+@Getter
+@Service
+public class LoginAttemptService {
+    private static final int STARTING_ATTEMPT = 0;
+    private static final int MAX_ATTEMPT = 3;
+    private static final int BLOCK_TIME_MILLIS = 300_000;
```
Comment (maximelyashevich) [inline src/main/java/com/epam/wca/gym/service/impl/LoginAttemptService.java:19]: Does it make sense to move such values to project properties?

File: src/main/java/com/epam/wca/gym/service/impl/TrainingServiceImpl.java [hunk 28]
```diff
@@ -5,13 +5,15 @@
...
 import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
 
 @Service
 @RequiredArgsConstructor
 public class TrainingServiceImpl implements TrainingService {
     private final TrainingRepository trainingRepository;
 
     @Override
+    @Transactional
```
Comment (maximelyashevich) [inline src/main/java/com/epam/wca/gym/service/impl/TrainingServiceImpl.java:16]: Why do we need `@Transactional` annotation for only one DB operation?

File: src/main/resources/application-dev.properties [hunk 29]
```diff
@@ -4,7 +4,7 @@ gym.api.server.description=Local Development Server
 
 # Password Configuration
 gym.api.profile.password.characters=ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789
```
Comment (maximelyashevich) [inline src/main/resources/application-dev.properties:6]: Could you please consider more secure solution to deal with sensitive details?

File: src/test/java/com/epam/wca/gym/controller/TrainingControllerTest.java [hunk 30]
```diff
@@ -1,5 +1,6 @@
 package com.epam.wca.gym.controller;
 
+// TODO: Simple http request test
 class TrainingControllerTest {
```
Comment (maximelyashevich) [inline src/test/java/com/epam/wca/gym/controller/TrainingControllerTest.java:4]: What is the value of this class?

File: src/test/java/com/epam/wca/gym/service/impl/UserDetailsServiceImplTest.java [hunk 31]
```diff
@@ -0,0 +1,80 @@
...
+
+import java.util.Optional;
+
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+class UserDetailsServiceImplTest {
```
Comment (maximelyashevich) [inline src/test/java/com/epam/wca/gym/service/impl/UserDetailsServiceImplTest.java:22]: **General comment:**
To improve your testing quality, please consider some useful articles on how to test security in Spring Boot applications:
- [Testing Method Security](https://docs.spring.io/spring-security/reference/servlet/test/method.html) (general overview)
- [Spring Security Testing](https://docs.spring.io/spring-security/site/docs/5.2.x/reference/html/test.html) (search by JWT)
- [Spring Security: Integration Tests](https://www.baeldung.com/spring-security-integration-tests) (practical examples)
- [Testing Security Configuration in Spring Boot](https://medium.com/@nagarjun_nagesh/testing-security-configuration-in-spring-boot-444b854bbd19) (practical examples)

The above materials are mostly for self-development.

**Review by maximelyashevich — APPROVED** [review]
Reviewed from my side, good job!

File: src/main/java/com/epam/wca/gym/aop/check/CheckTrainee.java
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/aop/check/CheckTrainee.java:1]: Done.

File: src/main/java/com/epam/wca/gym/aop/validation/TraineeValidator.java (hunk 22 above)
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/aop/validation/TraineeValidator.java:16]: Done.

File: src/main/java/com/epam/wca/gym/filter/JwtAuthenticationFilter.java (hunk 24 above)
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/filter/JwtAuthenticationFilter.java:36]: Yeah, it makes sense. Done.

File: src/main/java/com/epam/wca/gym/aop/validation/TraineeValidator.java (hunk 22 above)
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/aop/validation/TraineeValidator.java:16]: Done.

File: src/main/java/com/epam/wca/gym/filter/JwtAuthenticationFilter.java (hunk 24 above)
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/filter/JwtAuthenticationFilter.java:36]: Yeah, it does make sense. Done.

File: src/main/java/com/epam/wca/gym/monitoring/DatabaseInitMonitor.java (hunk 25 above)
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/monitoring/DatabaseInitMonitor.java:29]: Ok, done.

File: src/main/java/com/epam/wca/gym/monitoring/DatabaseInitMonitor.java (hunk 26 above)
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/monitoring/DatabaseInitMonitor.java:62]: Ok, done.

File: src/main/java/com/epam/wca/gym/service/impl/LoginAttemptService.java (hunk 27 above)
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/service/impl/LoginAttemptService.java:19]: Yeah, it does. Done.

File: src/main/java/com/epam/wca/gym/service/impl/TrainingServiceImpl.java (hunk 28 above)
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/service/impl/TrainingServiceImpl.java:16]: There were no real reason. Forgot about atomicity of operations. Removed.

File: src/test/java/com/epam/wca/gym/controller/TrainingControllerTest.java (hunk 30 above)
Comment (RYelzhan) [reply src/test/java/com/epam/wca/gym/controller/TrainingControllerTest.java:4]: To remember someday to finish this test. Should I remove it?

File: src/main/java/com/epam/wca/gym/aop/check/CheckTrainee.java
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/aop/check/CheckTrainee.java:1]: Done.

File: src/main/java/com/epam/wca/gym/aop/check/CheckTrainerAspect.java (hunk 21 above)
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/aop/check/CheckTrainerAspect.java:59]: Yeah, totally forgot about that. Done.

File: src/test/java/com/epam/wca/gym/service/impl/UserDetailsServiceImplTest.java (hunk 31 above)
Comment (RYelzhan) [reply src/test/java/com/epam/wca/gym/service/impl/UserDetailsServiceImplTest.java:22]: Ok, got it.

File: src/main/java/com/epam/wca/gym/controller/AuthenticationController.java (hunk 23 above)
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/controller/AuthenticationController.java]: Done.

File: src/main/java/com/epam/wca/gym/advice/RestAuthenticationEntryPoint.java (hunk 20 above)
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/advice/RestAuthenticationEntryPoint.java]: Yeah, done.

File: src/main/java/com/epam/wca/gym/config/SecurityConfig.java [hunk 32]
```diff
@@ -0,0 +1,98 @@
...
+        return http.build();
+    }
+
+    @Bean
+    public CorsFilter corsFilter() {
+        var source = new UrlBasedCorsConfigurationSource();
+        var config = new CorsConfiguration();
+
+        config.setAllowedOriginPatterns(Collections.singletonList("https://editor.swagger.io"));
+        config.setAllowedMethods(Collections.singletonList("*"));
```
[the remark this answers is not in the document]
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/config/SecurityConfig.java:77]: ok, got it, done.

**Pull request merged**